test/test_utils.h

```diff
@@ -159,23 +159,18 @@ bool compare_vectors(const std::vector<std::vector<T> *> &ref_vector,
 
     for (size_t i = 0; i < ref_vector.size(); i++) {
         for (size_t j = 0; j < ref_vector[i]->size(); j++) {
-            // We can't do the comparison directly because when the values are
-            // written and read from a file, they are converted to strings and
-            // back, which can cause some precision loss. So we convert them to
-            // strings and compare the strings.
-            std::stringstream aa;
-            aa << (*ref_vector[i])[j];
-            std::string a = aa.str();
-            std::stringstream bb;
-            bb << (*test_vector[i])[j];
-            std::string b = bb.str();
+            double test = (*test_vector[i])[j];
+            double ref = (*ref_vector[i])[j];
 
-            if (a != b) {
+            double diff = std::abs(test - ref);
```

Comment:
```suggestion
           double diff = std::abs(test - ref);
            double threshold =
                TOLERANCE * std::max(std::abs(test), std::abs(ref));

            if (diff > threshold) {
```
I think we need to take into account the magnitude. For example, assuming `test=1.23e-8` and `ref=1.4e-8`, therfore diff = `2e-9` which is smaller than `1e-8`. Comparing small values across different operating systems can be tough, but I prefer looking at the overall magnitude. I tested it on Ubuntu and MACOS, and it passed within the tolerance range. Still not entirely convinced though

Reply:
However, the test on CNN does not passed given that it is generated and tested on Ubuntu. Here is the output
```
Different values in cnn backward hidden states for mnist data
ref_vector[3][0] = -1.99999999e-12
test_vector[3][0] = -1.85488343e-12
diff 1.45116558e-13
threshold 7.99999997e-14
```
There might be an issue with CNN test. If you think there is no issue, we can hardcode line 167 following
```C++
if (diff > threshold && std::abs(ref) > 1e-10)
```
It means that we are not interested in those values smaller than 1e-10

Reply:
I still think that we have an issue with cnn-backward-hidden-states test